### PR TITLE
WIP - CHALS

### PR DESCRIPTION
**Pull request by degleris1**

### CODE REVIEW

File: cnmfpy/algs/chals.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    N, T = data.shape
+
+    # initial loss
+    model.loss_hist = [compute_loss(data, model.W, model.H)]
+
+    itr = 0
+    for itr in trange(model.n_iter_max):
+
+        if (np.isnan(model.W).any()):
+            raise Exception('W has Nans!!')
```
Comment (ahwillia) [inline cnmfpy/algs/chals.py]: nitpick -- `raise ValueError` instead of `raise Exception`

File: cnmfpy/algs/chals.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    # Set up Wk and residual
+    Wk = W[:, :, k].T  # TODO: will this slow things down?
+    norm_Wk = la.norm(Wk)
+    resid = X - tensor_conv(W, H)
+
+    # Update each entry in the row, from left to right
+    for t in range(T):
+        resid_slice = resid[:, t:t+L] + (H[k, t] * Wk)[:, :T-t]
+        H[k, t] = new_H_entry(Wk[:, :T-t], norm_Wk, resid_slice)
```
Comment (ahwillia) [inline cnmfpy/algs/chals.py]: After updating `H` we need to update `resid`. Imagine that `H[k, t]` goes from zero to one by this update. then `Wk` needs to be subtracted from the next `L` entries of resid.

File: cnmfpy/algs/chals.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        model.loss_hist += [compute_loss(data, model.W, model.H)]
+        model.W, model.H = renormalize(model.W, model.H)
+
+        # check convergence
+        prev_loss, new_loss = model.loss_hist[-2:]
+        if (np.abs(prev_loss - new_loss) < model.tol):
+            break
+
+
+def update_W_col(W, H, X, k, l):
```
Comment (ahwillia) [inline cnmfpy/algs/chals.py]: I'd vote to get rid of this function and just make it part of `update_W_component`

File: cnmfpy/algs/chals.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+def update_W_component(W, H, X, k):
+    """
+    """
+    L, N, K = W.shape
+
+    for l in range(L):
+        update_W_col(W, H, X, k, l)
+
+
+def new_H_entry(Wk, norm_Wk, resid_slice):
```
Comment (ahwillia) [inline cnmfpy/algs/chals.py]: Likewise, I'd put this into `update_H_component`

File: cnmfpy/algs/chals.py [hunk 5]
```diff
@@ -99,3 +99,4 @@ def update_H_component(W, H, X, k):
     for t in range(T):
         resid_slice = resid[:, t:t+L] + (H[k, t] * Wk)[:, :T-t]
```
Comment (bantin) [inline cnmfpy/algs/chals.py]: To save a few scalar-scalar multiplications, we could write this as `resid_slice = resid[:, t:t+L] + H[k, t] * Wk[:, :T-t]`, that way Numpy will only do the scalar mults for the columns we care about.

**Pull request merged**